test/gtest/possible-contents.cpp

```diff
@@ -0,0 +1,296 @@
+#include "ir/possible-contents.h"
+#include "wasm-s-parser.h"
+#include "wasm.h"
+#include "gtest/gtest.h"
+
+using namespace wasm;
+
+// Asserts a == b, in any order.
+template<typename T> void assertEqualSymmetric(const T& a, const T& b) {
+  std::cout << "\nassertEqualSymmetric\n";
+  a.dump(std::cout);
+  std::cout << '\n';
+  b.dump(std::cout);
+  std::cout << '\n';
+
+  EXPECT_TRUE(a == b);
+  EXPECT_TRUE(b == a);
```

Comment:
```suggestion
  EXPECT_EQ(a == b);
  EXPECT_EQ(b == a);
```
This will make gtest print a more helpful error message if the test fails. It would be good to keep the test output clean by removing the unconditional printing above and relying on gtest to do any necessary printing. If you implement `operator<<(std::ostream&)` to call `dump`, then gtest will find and use it automatically.

Reply:
I am a little unhappy about using `EXPECT_EQ` for a test that focuses on the implementation of the `==` operator itself. That is, `EXPECT_EQ` might be implemented internally as `assert(!(x != y))` - yes, that seems a little silly, but it could actually make sense for `EXPECT_NE` to not use `!=`. But my point is that it's best to make sure we actually see the `==` that is tested here.

I agree that using `operator<<` is useful. It would depend on the above point, however (if we need to keep the existing unconditional printing, then we don't need it).

Reply:
Thankfully the assertions are documented as testing the operators specifically: https://google.github.io/googletest/reference/assertions.html#EXPECT_EQ, so I don't think we need to worry about this. (Without that documentation I would agree that using EXPECT_TRUE would be better.)

Reply:
Ah, fair enough. To be honest even with such docs I am not happy depending on a third-party library that uses macros for something like this... but I suspect I'm overly paranoid here. I'll refactor in the more pragmatic way as you suggest.

Reply:
Although, wait... how would we check all the combinations? Atm we have this:
```cpp
  EXPECT_TRUE(a == b);
  EXPECT_FALSE(a != b);
```
The first line can be `EXPECT_EQ(a, b)`. But how can we write the second line? We want to write "not not equal" in order to test `operator!=`...

Reply:
Yep, we need to keep using `EXPECT_FALSE` for the second line.

Reply:
In that case is it worth changing stuff here? We'll still need the full manual logging for that case. The code would be more consistent to not switch between the "modes".

Reply:
It would be nice to get the better automatic error reporting where we can, even if we can't get it everywhere. And I hope we can avoid unconditionally printing anything.

But actually, looking at the gtest docs, it looks like what we want here is `EXPECT_PRED`, which lets us test an arbitrary condition and still get nice automatic printing: https://google.github.io/googletest/reference/assertions.html#EXPECT_PRED.